theano/tensor/nnet/nnet.py

```diff
@@ -1121,26 +1124,41 @@ def c_code(self, node, name, inp, out, sub):
                  "sm type should be float32 or float64");
             %(fail)s;
         }
-        if ((PyArray_NDIM(%(dnll)s) != 1)
+
+        // new scope because of variable declaration
+        // TODO: proper indentation, but the diff will get messy
+        {
+        // Get `dnll.shape[0]` or set it to zero if `dnll` is a scalar.
+        const npy_intp %(dnll)s_dims0 = (PyArray_NDIM(%(dnll)s) > 0 ?
+                                         PyArray_DIMS(%(dnll)s)[0] :
+                                         (npy_intp) 0);
+
+        // Get `dnll.strides[0]` and set it to zero if `dnll` is a scalar
+        // or a vector with just one element.
+        const npy_intp %(dnll)s_strides0 = (%(dnll)s_dims0 > 1 ?
+                                            PyArray_STRIDES(%(dnll)s)[0] :
+                                            (npy_intp) 0);
+
+        if ((PyArray_NDIM(%(dnll)s) > 1)
             || (PyArray_NDIM(%(sm)s) != 2)
             || (PyArray_NDIM(%(y_idx)s) != 1))
         {
             PyErr_SetString(PyExc_ValueError, "rank error");
             %(fail)s;
         }
-        if (PyArray_DIMS(%(dnll)s)[0] != PyArray_DIMS(%(sm)s)[0])
+        if (%(dnll)s_dims0 != PyArray_DIMS(%(sm)s)[0] && %(dnll)s_dims0 > 1)
```

Comment:
I don't agree because we only want to raise an error if the first dimensions of `%(dnll)s` and `%(sm)s` don't match or else if `%(dnll)s` is a "true vector", i.e. it has more than 1 element. I don't have to make a distinction between a scalar and a vector with only one element because I set `%(dnll)s_strides0` to zero in either case and, thus, the memory address in the loop will not be changed for the scalar-like cases because of `%(dnll)s_strides0 * i`. Do you agree?